plugins/modules/ec2_vpc_route_table.py

```diff
@@ -494,45 +534,88 @@ def ensure_subnet_association(connection=None, module=None, vpc_id=None, route_t
 
-    return {'changed': changed}
+    return changed
 
 
-def ensure_propagation(connection=None, module=None, route_table=None, propagating_vgw_ids=None,
-                       check_mode=None):
+def ensure_gateway_association(connection, module, route_table, gateway_id):
+    changed = False
+    if gateway_id:
+        # gateway_id specified - ensure route table is associated to gateway
+        filters = ansible_dict_to_boto3_filter_list({'association.gateway-id': gateway_id, 'vpc-id': route_table['VpcId']})
+        try:
+            route_tables = describe_route_tables_with_backoff(connection, Filters=filters)
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            module.fail_json_aws(e, msg="Couldn't get route tables")
+        for table in route_tables:
+            if table.get('RouteTableId'):
+                for association in table.get('Associations'):
+                    if association['Main']:
+                        continue
+                    if association.get('GatewayId') and association['GatewayId'] == gateway_id:
+                        if table['RouteTableId'] == route_table['RouteTableId']:
+                            return False
+                        if module.check_mode:
+                            return True
+                        try:
+                            connection.disassociate_route_table(
+                                aws_retry=True, AssociationId=association['RouteTableAssociationId'])
+                        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+                            module.fail_json_aws(e, msg="Couldn't disassociate gateway from route table")
+        try:
+            connection.associate_route_table(aws_retry=True,
+                                             RouteTableId=route_table['RouteTableId'],
+                                             GatewayId=gateway_id)
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            module.fail_json_aws(e, msg="Couldn't associate gateway with route table")
+        return True
+
+    else:
+        # No gateway_id specified - ensure route table is not associated to any gateways
+        to_delete = [association['RouteTableAssociationId'] for association in route_table['Associations'] if not association['Main']
+                     and association.get('GatewayId') and association['AssociationState']['State'] == 'associated']
+        for association_id in to_delete:
+            changed = True
+            if not module.check_mode:
+                try:
+                    connection.disassociate_route_table(aws_retry=True, AssociationId=association_id)
+                except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+                    module.fail_json_aws(e, msg="Couldn't disassociate gateway from route table")
+    return changed
```

Comment:
@jillr Not sure about this part here. But, I don't think we want to disassociate the the gw from the route_table if we do not specify one. I guess we should retain the previous configuration unless we explicitly want to disassociate it.

Reply:
I think we probably need to add a purge_gw parameter. Only when this parameter is true we want to disassociate the gw from the route table.

In any case, you should also handle the gw disassociation when state: absent.

Reply:
So I don’t disagree, but all I wanted to do here was add support for gateway route tables, so I tried to just replicate what was being done for subnet route tables and apply that to gateways. Subnets are dissociated when state is present and an empty array is passed in

Reply:
My only question is if a user specifies purge_gw as true and passes in the existing gw id, should we disassociate and re-associate the same gw, or do nothing?